### DISCUSSION TITLE
Похлебки Теклы

### DISCUSSION THREAD

**Discussion by dosinabox** [dosinabox/g2nr_unofficial_update#481] [Show and tell]
Мои наброски правок для похлебок Теклы:
[TheklaStews.zip](https://github.com/dosinabox/g2nr_unofficial_update/files/10147538/TheklaStews.zip)

Прошу прочитать, проанализировать и высказать мнение. В игре тоже можно протестировать наглядно. Эта переделка позволит перенести выдачу всех похлебок в один диалог и сделать похлебки за Буллко и гильдию независимыми друг от друга. В конечном итоге ГГ сможет получить **пять** похлебок без махинаций.

**Comment by Kor-Angar** [comment]
тут скорее условие касается 3 главы. в 4 такая реплика не возможна


```D
else if(Kapitel <= 4)
		{
			if(MIS_RescueBennet != LOG_SUCCESS)
			{
				AI_Output(self,other,"DIA_Thekla_PERM_17_14");	//Эти ублюдки из ополчения бросили Беннета за решетку.
				AI_Output(self,other,"DIA_Thekla_PERM_17_15");	//Окажи мне услугу, вызволи его оттуда, хорошо? А тем временем я сварю для тебя очень вкусную похлебку.
				if(MIS_RescueBennet != LOG_Running)
				{
					MIS_RescueBennet = LOG_Running;
					Log_CreateTopic(TOPIC_RescueBennet,LOG_MISSION);
					Log_SetTopicStatus(TOPIC_RescueBennet,LOG_Running);
					B_LogEntry(TOPIC_RescueBennet,"Кузнец Беннет арестован паладинами в городе.");
				};
			}
			else
			{
				AI_Output(self,other,"DIA_Thekla_PERM_17_16");	//У меня ничего не осталось. Заходи позже.
			};
		}
```

**Reply by dosinabox** [reply]
Условие на две главы:
- Глава 3: новость об аресте Беннета
- Глава 3 или 4: [после освобождения Беннета] выдача похлебки или фраза "У меня ничего не осталось".

**Comment by Kor-Angar** [comment]
так перекрывается квест с Беннетом. это не правильно. Беннет ключевое событие на ферме и в 3 главе это должно быть в приоритете. 

```D
	else if(MIS_Thekla_Paket == LOG_Running)
		{
			AI_Output(self,other,"DIA_Thekla_PERM_17_21");	//Нет трав - нет похлебки, понятно?
		}
		else if(Kapitel <= 2)
		{
			AI_Output(self,other,"DIA_Thekla_PERM_17_11");	//Эй, эй, эй! Не будь таким жадным! Я дам тебе знать, если у меня найдется еще работа для тебя.
			AI_Output(self,other,"DIA_Thekla_PERM_17_12");	//И ТОГДА ты получишь еще этой похлебки, понятно?
		}
		else if(Kapitel <= 4)
		{
			if(MIS_RescueBennet != LOG_SUCCESS)
```


я бы вообще диалог про Беннета сделал с `important`, при условие что Текла уже общалась с ГГ или ГГ наёмник.
```D
				AI_Output(self,other,"DIA_Thekla_PERM_17_14");	//Эти ублюдки из ополчения бросили Беннета за решетку.
				AI_Output(self,other,"DIA_Thekla_PERM_17_15");	//Окажи мне услугу, вызволи его оттуда, хорошо? А тем временем я сварю для тебя очень вкусную похлебку.
```

**Reply by dosinabox** [reply]
> так перекрывается квест с Беннетом

Фразой DIA_Thekla_PERM_17_21?

**Reply by Kor-Angar** [reply]
всеми вариантами что идут выше. то есть Текла вместо инфы про Беннета может дать квест про травы, а потом постоянно про травы спрашивать. в рамках твоей проверки Беннета поднят выше. вообще в целом лучше рассмотреть ситуацию что ГГ пришёл первый раз на ферму например с Миненталя. Беннет сидит в тюрьме. и вот при такой постановке смотреть как Текла будет себя вести.

**Reply by dosinabox** [reply]
> я бы вообще диалог про Беннета сделал с important, при условие что Текла уже общалась с ГГ или ГГ наёмник.

хорошая идея

**Reply by dosinabox** [reply]
А нужно ли это сообщать ополченцу?
```cpp
AI_Output(self,other,"DIA_Thekla_PERM_17_14");	//Эти ублюдки из ополчения бросили Беннета за решетку.
AI_Output(self,other,"DIA_Thekla_PERM_17_15");	//Окажи мне услугу, вызволи его оттуда, хорошо? А тем временем я сварю для тебя очень вкусную похлебку.
```

**Reply by Kor-Angar** [reply]
если Текла хорошо относится к герою, например он выполнил её "квест" про булко или сходил за травой то можно. если Текла видит героя мента или героя мага огня первый раз, то такое выдавать неправильно.

**Reply by Kor-Angar** [reply]
в `DIA_Thekla_BennetNews_Condition` можно добавить еще проверку на `C_TheklaCanGiveStew_Bullco`

**Reply by dosinabox** [reply]
тогда два important диалога будут конфликтовать друг с другом

**Reply by Kor-Angar** [reply]
3 глава. герой наёмник. булко побит, награда не получена. что будет происходить?

**Reply by dosinabox** [reply]
Сначала Текла даст похлебку за Буллко в important диалоге. При повторном обращении она расскажет про Беннета. Можно добавить проверку в диалог _EXIT и сообщать новость и там, чтобы диалоги не накладывались. Идеи?

**Reply by Kor-Angar** [reply]
а если попробовать так


```D
func int DIA_Thekla_AfterFight_Condition()
{
	if (C_TheklaCanGiveStew_Bullco() == true)
	&& (DIA_Thekla_BennetNews_Condition() == false)
	{
		return TRUE;
	};
};
```


```D
func void DIA_Thekla_BennetNews_Info()
{
	if (C_TheklaCanGiveStew_Bullco() == true)
	{
		DIA_Thekla_AfterFight_Info();
	};
	AI_Output(self,other,"DIA_Thekla_PERM_17_14");	//Эти ублюдки из ополчения бросили Беннета за решетку.
	AI_Output(self,other,"DIA_Thekla_PERM_17_15");	//Окажи мне услугу, вызволи его оттуда, хорошо? А тем временем я сварю для тебя очень вкусную похлебку.
	if(MIS_RescueBennet != LOG_Running)
	{
		MIS_RescueBennet = LOG_Running;
		Log_CreateTopic(TOPIC_RescueBennet,LOG_MISSION);
		Log_SetTopicStatus(TOPIC_RescueBennet,LOG_Running);
		B_LogEntry(TOPIC_RescueBennet,"Кузнец Беннет арестован паладинами в городе.");
	};
};
```

**Comment by Kor-Angar** [comment]
```D
		if(C_TheklaCanGiveStew_Guild())
		{
			if((other.guild == GIL_NOV) || (other.guild == GIL_KDF) || (other.guild == GIL_PAL))
			{
				AI_Output(self,other,"DIA_Thekla_Hunger_17_05");	//Как я могу отказать служителю Инноса?
			}
			else
			{
				AI_Output(self,other,"DIA_Thekla_Hunger_17_03");	//Вот твоя еда.
			};
			B_GiveInvItems(self,other,ItFo_XPStew,1);
			Thekla_GaveStew_Guild = TRUE;
		}
```
вот твоя еда доступно только Наёмнику и ОнД?

**Reply by dosinabox** [reply]
Да. Фразы для нейтрала и ополченца ниже.
Ну и для прочих экзотических гильдий тоже "Вот твоя еда", например если ГГ вдруг станет GIL_KDW.

**Reply by Kor-Angar** [reply]
мент приходит первый раз на ферму во второй главе. говорит с ней. судя по коду она даст ему квест про Сагиту? :-)

**Reply by dosinabox** [reply]
Даст квест ополченцу только если он подерется с Буллко/Сильвио. В оригинале так же.

**Reply by Kor-Angar** [reply]
представим что есть квест про побить Булко. хотя там переменной нет, но допустим она есть. тогда этот квест герой "получает" от Теклы будучи нейтралом в первой главе. при этом "выполнить" этот квест герой может в 1-3 главе принадлежа любой фракции. то есть это общая канва готика, набираем квесты, сдаём по мере возможностей. не взял квест нейтралом, не сможешь его потом сдать.

тут я как понимая схожая история хотя переменной на квест и нет. по сути данный "квест" взять может только нейтрал после первого отказа в получение супа.

мент за избиение Булко при каких условиях получает суп?

**Reply by dosinabox** [reply]
Квеста на драку с Буллко от Теклы нет, его не надо ни брать, ни сдавать. Нужно просто подраться с Буллко или Сильвио, пока они не ушли в Миненталь. Выигрывать не обязательно, достаточно получить нокаут от Буллко или бессмертного Сильвио.
Сразу после драки с любым исходом Текла дает свой комментарий и похлебку. Данный диалог ограничен четвертой главой, когда он перестает быть актуальным (если подраться в 1-3 главе, уйти в Миненталь, начать четвертую главу и вернуться, то похлебки не будет).

> мент за избиение Булко при каких условиях получает суп?

То же самое. Сильвио побил ГГ / Буллко побил ГГ / ГГ побил Буллко. А в четвертой главе ГГ уже не может быть ополченцем.

**Reply by Kor-Angar** [reply]
понятно.

**Comment by Kor-Angar** [comment]
про пакет трав я так не скажу. надо в игре пробовать. возможно было бы лаконично его автоматически выдавыть в диалоге про награду за беннета и булко

**Reply by Kor-Angar** [reply]
а почему ты не хочешь давать суп прям в том же диалоге где герой отдаёт пакет `DIA_Thekla_SagittaPaket_Info` ?  можно там и на месте приготовить что то как это делает повар в ярике.

**Reply by dosinabox** [reply]
Можно и давать, это разгрузит основной диалог. Но какие-то реплики не очень подходящие. Не знаю.

**Reply by Kor-Angar** [reply]
```D
	AI_Output(other,self,"DIA_Thekla_SagittaPaket_15_00");	//Вот пакет от Сагитты.
	AI_Output(self,other,"DIA_Thekla_SagittaPaket_17_01");	//Огромное спасибо. От тебя есть хоть какая-то польза в отличие от других.
	AI_Output(self,other,"DIA_Thekla_Hunger_17_03");	//Вот твоя еда.
```




```D
	AI_Output(other,self,"DIA_Thekla_SagittaPaket_15_00");	//Вот пакет от Сагитты.
	AI_Output(self,other,"DIA_Thekla_SagittaPaket_17_01");	//Огромное спасибо. От тебя есть хоть какая-то польза в отличие от других.
	AI_Output(other,self,"DIA_Thekla_Hunger_15_00");	//Я голоден!
	AI_Output(self,other,"DIA_Thekla_PERM_17_10");	//Хорошо. Я сжалюсь над тобой. Вот, держи. Не могу смотреть, как ты умираешь от голода у меня на глазах.
```



```D
	AI_Output(other,self,"DIA_Thekla_SagittaPaket_15_00");	//Вот пакет от Сагитты.
	AI_Output(self,other,"DIA_Thekla_SagittaPaket_17_01");	//Огромное спасибо. От тебя есть хоть какая-то польза в отличие от других.
	AI_Output(other,self,"DIA_Thekla_PERM_15_04");	//А могу я облизать горшок, а?
	AI_Output(self,other,"DIA_Thekla_PERM_17_10");	//Хорошо. Я сжалюсь над тобой. Вот, держи. Не могу смотреть, как ты умираешь от голода у меня на глазах.
```

**Reply by dosinabox** [reply]
Я не люблю переиспользовать реплики или растаскивать их по разным диалогам)
```cpp
AI_Output(other,self,"DIA_Thekla_SagittaPaket_15_00");	//Вот пакет от Сагитты.
AI_Output(self,other,"DIA_Thekla_SagittaPaket_17_01");	//Огромное спасибо. От тебя есть хоть какая-то польза в отличие от других.
AI_Output(other,self,"DIA_Bennet_Present_15_14");	//И?
AI_Output(self,other,"DIA_Thekla_PERM_17_10");	//Хорошо. Я сжалюсь над тобой. Вот, держи. Не могу смотреть, как ты умираешь от голода у меня на глазах.
```

**Reply by Kor-Angar** [reply]
> Я не люблю переиспользовать реплики или растаскивать их по разным диалогам)

это мучение :-)

**Comment by Kor-Angar** [comment]
> Квеста на драку с Буллко от Теклы нет, его не надо ни брать, ни сдавать. Нужно просто подраться с Буллко или Сильвио, пока они не ушли в Миненталь. Выигрывать не обязательно, достаточно получить нокаут от Буллко или бессмертного Сильвио.
> Сразу после драки с любым исходом Текла дает свой комментарий и похлебку. Данный диалог ограничен четвертой главой, когда он перестает быть актуальным (если подраться в 1-3 главе, уйти в Миненталь, начать четвертую главу и вернуться, то похлебки не будет).

я пересмотрел диалоги и условия про "**хорошие манеры**" и у теклы и у Сильвио 
`DIA_Thekla_WannaJoin` --> `DIA_Thekla_Problem `--> `DIA_Thekla_Manieren`

и.. я упорно(упорото) не понимаю почему это не квест. то есть это полноценный квест со всеми признаками квеста 4 или 5 диалогов с взаимными проверками и награда :-D 
из диалога награды чётко видно что до этого у них был разговор. реплика DIA_Thekla_AfterFight_17_00 и DIA_Thekla_AfterFight_17_03 явно на это указывают :-)
```D
func void DIA_Thekla_AfterFight_Info ()
{
	if (Bullco.aivar[AIV_LastFightAgainstPlayer] == FIGHT_LOST)
	{
		AI_Output (self, other, "DIA_Thekla_AfterFight_17_00"); //Du hast es Bullco aber ganz schцn gezeigt, Schдtzchen.
		AI_Output (self, other, "DIA_Thekla_AfterFight_17_01"); //War bestimmt anstrengend, das fette Schwein zu verprьgeln.
	}
	else //Sylvio oder Bullco gewonnen
	{
		AI_Output (self, other, "DIA_Thekla_AfterFight_17_02"); //Du hast dich ja ganz schцn verprьgeln lassen, Schдtzchen.
		AI_Output (self, other, "DIA_Thekla_AfterFight_17_03"); //Hab ich's dir nicht gesagt? Jetzt weiЯt du, was ich meine.
		AI_Output (other, self, "DIA_Thekla_AfterFight_15_04"); //Freut mich, wenn du dich amьsierst.
		AI_Output (self, other, "DIA_Thekla_AfterFight_17_05"); //Jetzt zieh mal nicht so'n Gesicht. Du bist nicht der Erste, der von dem Abschaum verprьgelt wurde.
	};
	
	AI_Output (self, other, "DIA_Thekla_AfterFight_17_06"); //Hier, iss dir erst mal 'n Happen, damit du wieder zu Krдften kommst.
	B_GiveInvItems (self, other, ItFo_XPStew, 1);	
	Thekla_GaveStew = TRUE;
};
```
в целом я понимаю что ты просто хочешь причесать код, не меняя структуры. но если свободное время еще раз присмотрись к этому моменту.
вполне может быть что это был квест который просто недооформили или не повесили доп условия который брал нейтрал не получивший суп за гильдию. то есть может быть что супчиков максимум вовсе не 5, а 4 :-D

**Comment by Kor-Angar** [comment]
обновлённый файл будет? :-)

**Reply by dosinabox** [reply]
- новость об аресте Беннета вынесена в отдельный диалог
- для похлебки за освобождение Беннета не нужно приносить травы от Сагитты
- выдача похлебки за освобождение Беннета поднята выше похлебки за травы от Сагитты
- правка диалога DIA_Thekla_SagittaPaket под текущие наработки
- Текла не спросит наемника и ОнД что он делает на её кухне

[TheklaStews_v2.zip](https://github.com/dosinabox/g2nr_unofficial_update/files/10148533/TheklaStews_v2.zip)

**Reply by Kor-Angar** [reply]
всё посмотрел. дальше нужно тестить :-)

**Comment by Kor-Angar** [comment]
в `DIA_Thekla_PERM_Condition` и  `DIA_Thekla_PERM_Info` можно использовать проверку 	`if (Npc_KnowsInfo(hero,DIA_Thekla_PERM)` это я просто напоминаю периодически про это. вдруг когда нибудь такой метод тебе подойдёт :-)

**Reply by Kor-Angar** [reply]
проверку на` (Npc_KnowsInfo(hero,DIA_Thekla_BennetNews)`  можно использовать в  `DIA_Thekla_BennetNews_Info `

**Reply by dosinabox** [reply]
DIA_Thekla_PERM (а теперь и DIA_Thekla_BennetNews) - `permanent`, проверять такие диалоги можно только с Union.

**Reply by Kor-Angar** [reply]
> проверять такие диалоги можно только с Union.

не очень понял. что имеется ввиду?

**Reply by Kor-Angar** [reply]
без юнион он не делал запипси о том что инфо функции инстанций с `permanent = true`  уже вызывались? 
или `Npc_KnowsInfo` работало некорректно с инстанциями у которых `permanent = true` ?

**Reply by dosinabox** [reply]
https://worldofplayers.ru/threads/36995/post-1154539

**Reply by Kor-Angar** [reply]
секретные материалы :-)
![image](https://user-images.githubusercontent.com/49514764/205725983-57052878-e928-4600-8641-db9c3e09739f.png)

**Reply by dosinabox** [reply]
> D36:
Диалоги со свойством permanent = 0 проигрываются однократно. Диалоги со свойством permanent = 1 доступны всё время, пока выполняются их условия (например, опции алхимии или диалоги обучения). Если ГГ прослушал такой диалог, то этот факт нельзя проверить с помощью Npc_KnowsInfo, такая проверка всегда будет возвращать FALSE. Например, в Г2НВ недоступны диалоги DIA_Gunnar_South, DIA_Gunnar_Colony и DIA_Gunnar_Bandits, потому что все проверяют перманентный диалог DIA_Gunnar_Everywhere.

> N1kX:
Примечательно, что в оффкоменте экстернала функции об этом написано
FUNC INT Npc_KnowsInfo (VAR C_NPC self, VAR INT infoInstance) { return 0; };
// Возвращает TRUE, если данный игрок уже получил информацию.
// ВНИМАНИЕ: даже если игроку уже была сообщена постоянная информация, эта функция все равно возвращает FALSE!

> Gratt:
Если верить патчу, он как раз должен после фразы DIA_Gunnar_Everywhere разблокировать DIA_Gunnar_South. Так как теперь функция Npc_KnowsInfo будет возвращать результат игнорируя свойство перманентности.

**Reply by Kor-Angar** [reply]
понятно

**Reply by Kor-Angar** [reply]
а я всё равно буду напоминать :-) в "Blood Night / Ночь крови"  можно использовать проверку на if (Npc_KnowsInfo(hero,DIA_Lutero_EXIT)

```D
func void DIA_Lutero_EXIT_Info()
{
	if((FIRST_TIME == FALSE) && (other.guild != GIL_KDF))
	{
		AI_Output(self,other,"Dia_Luttero_wolfe_00");	//Подожди! У меня есть предложение, от которого ты не сможешь отказаться.
		AI_Output(other,self,"Dia_Luttero_wolfe_01");	//Чего ради?
		AI_Output(self,other,"Dia_Luttero_wolfe_02");	//Один мой клиент захотел шубу, но не простую: он хочет сшить ее из шкур черных волков.
		AI_Output(other,self,"Dia_Luttero_wolfe_03");	//Черных волков?
		AI_Output(self,other,"Dia_Luttero_wolfe_04");	//Ага, черных волков. Их мех особенно ценится, потому что это вымирающий вид. Но поговаривают, что на Хоринисе еще обитает несколько.
		AI_Output(self,other,"Dia_Luttero_wolfe_05");	//Если ты принесешь мне шкуры трех черных волков, я щедро награжу тебя.
		AI_Output(other,self,"Dia_Luttero_wolfe_06");	//Я посмотрю, что можно сделать.
		Log_CreateTopic(SCHWARZE_WOELFE,LOG_MISSION);
		Log_SetTopicStatus(SCHWARZE_WOELFE,LOG_Running);
		B_LogEntry(SCHWARZE_WOELFE,"Торговец Лютеро заказал мне три шкуры черных волков.");
		FIRST_TIME = TRUE;
	};
```

**Comment by dosinabox** [comment]
- выдача похлебки за травы происходит при сдаче трав
- выдача похлебки за Беннета происходит отдельно
- новость об аресте Бенната и выдача похлебки после освобождения - один диалог DIA_Thekla_BennetNews
- диалог о Беннете DIA_Thekla_BennetNews на всякий случай дублируется в DIA_Thekla_EXIT

[TheklaStews_v3.zip](https://github.com/dosinabox/g2nr_unofficial_update/files/10153799/TheklaStews_v3.zip)

**Reply by Kor-Angar** [reply]
вводится `Thekla_GaveStew_Sagitta ` просто для единообразия?

```D
	Thekla_GaveStew_Sagitta = TRUE;
	MIS_Thekla_Paket = LOG_SUCCESS;
```

**Reply by Kor-Angar** [reply]
посмотрел. ничего такого не увидел больше. за что то глаз цепляется, но это просто из области проверять проверки местами, что в конечном итоге ни на что не влияет. надо тестить в игре. скинь файл сагитты и тех кого затрагивают изменения. или сделай свежий vdf том со скриптпатч для тестов.

**Reply by dosinabox** [reply]
> вводится `Thekla_GaveStew_Sagitta ` просто для единообразия?

Нужно для подсчета общего количества полученных похлебок TheklaStewsGiven(). И потом можно использовать для статистики.

> скинь файл сагитты и тех кого затрагивают изменения

Выше, в TheklaStews_v3.zip. Этого достаточно, должно работать и с 28 версией и с версией из мастерской. Я так и тестирую, только Юнион нужен.

**Comment by dosinabox** [comment]
- новости о Беннете на всякий случай добавлены и в общий диалог DIA_Thekla_PERM

[TheklaStews_v4.zip](https://github.com/dosinabox/g2nr_unofficial_update/files/10158200/TheklaStews_v4.zip)

**Comment by dosinabox** [comment]
Текущая логика:
- если убить Буллко, то Текла не даст похлебку
- есть убить Буллко и дать Сильвио себя побить, то Текла даст похлебку

Всё нормально?)

**Reply by Kor-Angar** [reply]
по идеи должен быть отдельный вариант при смерти Булко с провалом миссии. но его в игре нет. реплик нет. что то можно попробовать конечно сочинить.
